src/main/kotlin/com/cultureamp/eventsourcing/RelationalDatabaseEventStore.kt

```diff
@@ -256,36 +247,57 @@ object H2DatabaseEventStore {
         noinline afterSinkHook: (List<SequencedEvent<M>>) -> Unit,
         eventsSinkTableName: String,
     ): RelationalDatabaseEventStore<M> {
-        return RelationalDatabaseEventStore(db, eventsTable(tableName), EventsSequenceStats(eventsSequenceStateTableName), M::class.java, objectMapper, eventTypeResolver, afterSinkHook = afterSinkHook, eventsSinkTable = Events(eventsSinkTableName, Table::jsonb), )
+        return RelationalDatabaseEventStore(
+            db,
+            Events(tableName, objectMapper, object : TypeReference<M>() {}),
+            EventsSequenceStats(eventsSequenceStateTableName),
+            objectMapper,
+            eventTypeResolver,
+            afterSinkHook = afterSinkHook,
+            eventsSinkTable = Events(eventsSinkTableName, objectMapper, object : TypeReference<M>() {}),
+        )
     }
-
-    @PublishedApi
-    internal fun eventsTable(tableName: String = defaultEventsTableName) = Events(tableName) { name -> this.text(name) }
 }
 
 internal fun <T> String.asClass(): Class<out T>? {
     @Suppress("UNCHECKED_CAST")
     return Class.forName(this) as Class<out T>?
 }
 
-class Events(tableName: String = defaultEventsTableName, jsonb: Table.(String) -> Column<String> = Table::jsonb) :
-    Table(tableName) {
+class Events<M : EventMetadata>(tableName: String = defaultEventsTableName, private val objectMapper: ObjectMapper = defaultObjectMapper, metadataType: TypeReference<M>) : Table(tableName) {
+    private val mapType = object : TypeReference<Map<String, Any>>() {}
+
     val sequence = long("sequence").autoIncrement()
     val eventId = uuid("id")
     val aggregateSequence = long("aggregate_sequence")
     val aggregateId = uuid("aggregate_id")
     val aggregateType = varchar("aggregate_type", 128)
     val eventType = varchar("event_type", 256)
     val createdAt = datetime("created_at")
-    val body = jsonb("json_body")
-    val metadata = jsonb("metadata")
+    val body = jsonb("json_body", { validatedSerialize(it, mapType) }, { deserialize(it, mapType) })
```

Comment:
is there some reason we don't use the jackson-based kotlinx serialization? https://jetbrains.github.io/Exposed/data-types.html#how-to-use-json-and-jsonb-types

Reply:
kotlinx serialization is different from jackson. I actually prefer kotlinx serialization because it gives you compile-time serialization checking. But it's actually a huge API change - it means all users of Kestrel will need to change their event classes to all implement `@Serializable` - and it may lead to subtle changes that make the kotlinx generated json different to the jackson json, which will mean existing data in the table may not deserialize propely. But also in this case, it still wouldn't help us because of how we store the event-type in a different column, so we'd need to do some mucking around either way